### PR TITLE
Return warmup request error during model loading

### PR DESCRIPTION
**Pull request by GuanLuo**

### CODE REVIEW

File: src/backend_model_instance.cc
Comment (rmccorm4) [inline src/backend_model_instance.cc:582]: Can you write a comment here summarizing this process? 

> ... set response callback for each warmup request and pass response_complete userp so we can check for errors from each async response and exit early if any warmup request fails ...

> ... use promises to signal when each response is complete async ... etc.

This part of the code is new to me so these kinds of summaries are very helpful when reading it later.

Otherwise, the WIP LGTM so far

**Review by rmccorm4 — CHANGES_REQUESTED** [review]

**Review by Tabrizian — COMMENTED** [review]
Do we want to add some testing for decoupled models using warmup feature too?

**Comment by GuanLuo** [comment]
@tanmayv25 Do you think if this [commit](https://github.com/triton-inference-server/core/pull/77/commits/3c1893102080d6b4eced389e713789603b08e252) makes sense? I am thinking about warmup for sequence model where the batch will be formed by batch-1 requests in regular inference. This is the same before the commit, after the commit the control inputs are "concatenated" to a single input from the backend's perspective.

I think it is okay though as the backend doesn't examine the control input of the requests anyway, it treats them like regular input: put them into single buffer and send to the model.

File: src/backend_model_instance.cc
Comment (rmccorm4) [inline src/backend_model_instance.cc:449]: Why is a string treated differently?

Comment (GuanLuo) [reply src/backend_model_instance.cc:452]: Added comment

**Review by rmccorm4 — COMMENTED** [review]
LGTM, though I can't confidently comment on correctness of this comment: https://github.com/triton-inference-server/core/pull/77#issuecomment-1129238067

I'll leave that to @tanmayv25

**Comment by tanmayv25** [comment]
Some of the backends may be using control inputs to identify/manage state within the backend. I would prefer if we don't merge the control inputs.

**Comment by GuanLuo** [comment]
Reverted single request change

**Review by rmccorm4 — APPROVED** [review]

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**
